numpy/linalg/tests/test_linalg.py

```diff
@@ -923,6 +919,42 @@ def test_future_rcond(self):
             # Warning should be raised exactly once (first command)
             assert_(len(w) == 1)
 
+    @pytest.mark.parametrize(["m", "n", "n_rhs"], [
+        (4, 2, 2),
+        (0, 4, 1),
+        (0, 4, 2),
+        (4, 0, 1),
+        (4, 0, 2),
+        (4, 2, 0),
+        (0, 0, 0)
+    ])
+    def test_empty_a_b(self, m, n, n_rhs):
+        a = np.arange(m * n).reshape(m, n)
+        b = np.ones((m, n_rhs))
+        x, residuals, rank, s = linalg.lstsq(a, b, rcond=None)
+        assert_equal(x.shape, (n, n_rhs))
+        assert_equal(residuals.shape, ((n_rhs,) if m > n else (0,)))
+        if m > n and n_rhs > 0:
+            # residuals are exactly the squared norms of b's columns
+            r = b - np.dot(a, x)
+            assert_almost_equal(residuals, (r * r).sum(axis=-2))
+        assert_equal(rank, min(m, n))
+        assert_equal(s.shape, (min(m, n),))
```

Comment:
How much of this block could be replaced with `self.do`? It seems there's a fair amount of repetition between this test and the one that was used before

Reply:
I think that part can be left to another enhancement.

Because I would have to add test cases that apply to everything. That's fine, and it would be great.

`cond` (around L1566) still has `_assertNoEmpty2d`... and we can just have `cond(m)=0` for empty matrix `m` (because sup_x (|Ax| / |x|) = 0 when |Ax| = 0 for all x).

Currently, the `qr` code is in master but not in this branch. I could cherry pick it over, but I'm not sure if things might get ugly.

Once the two are in, we won't have any more `_assertNoEmpty2d`'s running about.

Reply:
The whole `test_empty_a_b_1d` bit can be exactly performed with `self.do` but as for the other, `test_empty_a_b` has `b` which is 2d. So I'd keep that.

Reply:
I'd leave `cond` as is for now. Keeping `test_empty_a_b` sounds fine to me - just wanted to make sure you'd considered using `.do`. Switching to using it or hearing an argument why you didn't (`b` being 2d) are both good responses.